### ISSUE TITLE
The perlin_interp function not returning correct values?

### ISSUE THREAD

**Issue by 40163650** [petershirley/raytracingthenextweek#6]
When I followed the tutorial in the book, I found that just before the turbulence is added I encountered some issues with my scene not rendering correctly.

I expected it to look like this:
![screenshot_1_expected](https://user-images.githubusercontent.com/12065481/48987023-f3652e80-f112-11e8-82cf-0a1d10cc07e6.png)
What I actually got was this:
![screenshot_1_actual](https://user-images.githubusercontent.com/12065481/48987032-00821d80-f113-11e8-9ecb-a9ae5708ae7f.png)

After continuing the tutorial to the turbulence section, I noticed that fabs() was used on the accumulated total. I wondered if this would fix my issue with the perlin_interp function (i.e. noise.noise() not noise.turb()), so I used it at the end there.

It did get rid of the black boxes but the result was also still not quite what was intended:
![screenshot_1](https://user-images.githubusercontent.com/12065481/48987116-d3823a80-f113-11e8-9456-ddf266fe197a.png)

Perhaps this is because I used a scale value of 1, but setting it higher leads to results more similar to using turbulence.

**Comment by damonbla** [comment]
Can confirm. I'm getting the same exact results at this point of the book. And the same when using fabs(). Unfortunately I'm not sure how to fix it! I can at least say I had the same looking images.

**Comment by 40163650** [comment]
Update: I left this issue for a while, and progressed with the rest of the book but I went back to it today.
I decided to clamp the value before outputting it to the texture file instead of in the perlin_interp() function and got a result that looked closer to the intended result, but it still had a lot of contrast. Here it is:

![screenshot_1](https://user-images.githubusercontent.com/12065481/49609519-ac3d3000-f993-11e8-9921-4046a741099c.png)

**Comment by obiwanus** [comment]
I think that the low contrast image provided in the book is not what you'll receive if you follow along (there have been errors in the book up to that point, so it might just be one of them).

If you apply `turb` using the perlin noise function that gives the high contrast result as above, you will get the expected output just like in the book.

**Comment by tylermorganwall** [comment]
The issue with the book version is `perlin_interp`, without the `turb` function, can return negative values due to the dot product. This leads to negative values being passed to the sqrt function. You can recreate the correct output at that specific point in the book simply by rescaling the dot product from -1 to 1 to 0 to 1.

```c
inline float perlin_interp(vec3 c[2][2][2], float u, float v, float w) {
  float uu = u*u*(3-2*u);
  float vv = v*v*(3-2*v);
  float ww = w*w*(3-2*w);
  float accum = 0;
  for(int i = 0; i < 2; i++) {
    for(int j = 0; j < 2; j++) {
      for(int k = 0; k < 2; k++) {
        vec3 weight_v(u-i,v-j, w-k);
        accum += (i*uu + (1-i)*(1-uu))*
                 (j*vv + (1-j)*(1-vv))*
                 (k*ww + (1-k)*(1-ww))*
                 0.5 * (1 + dot(c[i][j][k], weight_v));
      } 
    }
  }
  return(accum);
}
```

**Comment by fralken** [comment]
@tylermorganwall you are right. An alternative can be to scale the final `accum` result, instead of each  dot product:

```c
inline float perlin_interp(vec3 c[2][2][2], float u, float v, float w) {
  float uu = u*u*(3-2*u);
  float vv = v*v*(3-2*v);
  float ww = w*w*(3-2*w);
  float accum = 0;
  for(int i = 0; i < 2; i++) {
    for(int j = 0; j < 2; j++) {
      for(int k = 0; k < 2; k++) {
        vec3 weight_v(u-i,v-j, w-k);
        accum += (i*uu + (1-i)*(1-uu))*
                 (j*vv + (1-j)*(1-vv))*
                 (k*ww + (1-k)*(1-ww))*dot(c[i][j][k], weight_v);
      } 
    }
  }
  return (accum + 1.0) * 0.5;
}
```
I think it brings to similar results
